### PR TITLE
coral-web: share conversations

### PR DESCRIPTION
**Pull request by misspia-cohere**
Porting share links from Coral with addition of tool events + citations

## No citations desktop + mobile

https://github.com/cohere-ai/cohere-toolkit/assets/140425731/af8a9380-62c2-4322-a04b-3121854c4a02

## With citations desktop + mobile

https://github.com/cohere-ai/cohere-toolkit/assets/140425731/b1df033a-0e0e-4185-969f-c1ac9fd4f2a8

**AI Description**

This PR introduces a new feature that allows users to create, view, and share snapshots of conversations. It also adds a "Share" option to the conversation header menu, enabling users to generate and share links to their conversations.

## Files Changed
- `src/interfaces/coral_web/src/cohere-client/client.ts`: Adds snapshot-related methods to the `CohereClient` class.
- `src/interfaces/coral_web/src/cohere-client/generated/schemas.gen.ts`: Updates the `$Snapshot` type and adds new types for snapshot-related data structures.
- `src/interfaces/coral_web/src/cohere-client/generated/services.gen.ts`: Adds a method to search conversations by title.
- `src/interfaces/coral_web/src/cohere-client/generated/types.gen.ts`: Updates existing types and adds new types for snapshot-related data structures.
- `src/interfaces/coral_web/src/components/Conversation/Header.tsx`: Adds a "Share" option to the conversation header menu and a `ShareModal` component.
- `src/interfaces/coral_web/src/components/ShareModal.tsx`: Implements the `ShareModal` component, which allows users to generate and share links to their conversations.
- `src/interfaces/coral_web/src/hooks/snapshots.ts`: Adds hooks for working with snapshots, including creating, listing, and retrieving them.
- `src/interfaces/coral_web/src/pages/share/[id]/index.tsx`: Implements the "Share Conversation" page, allowing users to view and share snapshots of conversations.
- `src/interfaces/coral_web/src/utils/error.ts`: Adds a function to check if a share link has expired.
- `src/interfaces/coral_web/tailwind.config.js`: Updates the "share-content" and "share-content-with-citations" widths in the Tailwind configuration.

## Summary
- **New Feature**: Snapshot creation and sharing.
- **New Component**: `ShareModal`.
- **Updated Components**: `Header`, `ReadOnlyConversation`.
- **New Hooks**: `useCreateSnapshotLinkId`, `useListSnapshots`, `useGetSnapshotByLinkId`, `useSnapshots`.
- **New Page**: "Share Conversation".

### CODE REVIEW

File: src/interfaces/coral_web/src/components/ReadOnlyConversation.tsx [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+
+type ReadOnlyConversationCitationProps = {
+  generationId: string;
+  message: string;
+};
+
+/**
+ * @description Citation for read only view of a shared conversation with simplified positioning
+ */
+const ReadOnlyConversationCitation: React.FC<ReadOnlyConversationCitationProps> = ({
```
Comment (misspia-cohere) [inline src/interfaces/coral_web/src/components/ReadOnlyConversation.tsx:101]: Simplified version of `Citation.tsx`
- Aligns with the top of the message row
- Does not jump next to the selected highlighted text in the message

**Comment by lusmoura** [comment]
Is this a bug in the response?

image

**Comment by misspia-cohere** [comment]
@lusmoura seems to happen whenever there is an incomplete response from the bot (ie. response is aborted mid generation. might happen with error'd out responses but unable to confirm atm)-- if you want to test in the UI, you can hit the square/stop button in the composer while the bot is streaming in a response

### Conversation
Screenshot 2024-07-11 at 12 43 54

### Share view
Screenshot 2024-07-11 at 12 43 39

File: src/interfaces/coral_web/src/components/ReadOnlyConversation.tsx [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+        </Text>
+
+        <div className={cn('my-6 w-full border-b border-marble-500')} />
+
+        <div className="flex flex-col gap-y-4 py-6 md:gap-y-6">
+          {messages.map((m, i) => (
+            <div key={i} className="flex items-start gap-x-3">
+              <MessageRow
+                message={m}
+                isLast={messages.length - 1 === i}
```
Comment (wujessica) [inline src/interfaces/coral_web/src/components/ReadOnlyConversation.tsx:75]:
```suggestion
                isLast={i === messages.length - 1}
```

File: src/interfaces/coral_web/src/components/ReadOnlyConversation.tsx (hunk 1 above)
Comment (wujessica) [reply src/interfaces/coral_web/src/components/ReadOnlyConversation.tsx:101]: feel free to put this in the description

File: src/interfaces/coral_web/src/components/ShareModal.tsx [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+    } else {
+      const generateNewSnapshotLink = async () => {
+        setStatus('modal-loading');
+        try {
+          const linkId = await createSnapshotLinkId({ conversationId });
+          if (linkId) {
+            setLinkId(linkId);
+            setStatus(undefined);
+          } else {
+            setStatus('modal-error');
```
Comment (wujessica) [inline src/interfaces/coral_web/src/components/ShareModal.tsx:41]: when does this happen?

File: src/interfaces/coral_web/src/hooks/snapshots.ts [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+const useListSnapshots = () => {
+  const client = useCohereClient();
+
+  return useQuery<ListSnapshotsV1SnapshotsGetResponse, Error>({
+    queryKey: ['listSnapshots'],
+    queryFn: async () => {
+      try {
+        return await client.listSnapshots();
+      } catch (e) {
+        throw e;
```
Comment (wujessica) [inline src/interfaces/coral_web/src/hooks/snapshots.ts:28]: to be consistent with the rest of this file:
```suggestion
        console.error(e);
        throw e;
```

File: src/interfaces/coral_web/src/hooks/snapshots.ts [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+export const useCreateSnapshotLinkId = () => {
+  const client = useCohereClient();
+  const queryClient = useQueryClient();
+  return useMutation({
+    mutationFn: async ({ conversationId }: { conversationId: string }) => {
+      try {
+        const newSnapshot = await client.createSnapshot({ conversation_id: conversationId });
+        return newSnapshot.link_id;
+      } catch (e) {
+        console.error(e);
```
Comment (wujessica) [inline src/interfaces/coral_web/src/hooks/snapshots.ts:48]:
```suggestion
        console.error(e);
        throw e;
```

File: src/interfaces/coral_web/src/hooks/snapshots.ts [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+  const getSnapshotLinksByConversationId = (conversationId: string) =>
+    snapshots
+      .filter((s) => s.conversation_id === conversationId)
+      .map((s) => s.links)
+      .flat();
+
+  const deleteAllSnapshotLinks = async (conversationId: string): Promise<void> => {
+    const snapshot = snapshots.find((s) => s.conversation_id === conversationId);
+    if (!snapshot) return;
+    if (!snapshot.links) return;
```
Comment (wujessica) [inline src/interfaces/coral_web/src/hooks/snapshots.ts:89]:
```suggestion
    if (!snapshot || !snapshot.links) return;
```

File: src/interfaces/coral_web/src/hooks/snapshots.ts [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+    try {
+      await Promise.all(
+        snapshot?.links?.map(async (linkId) => {
```
Comment (wujessica) [inline src/interfaces/coral_web/src/hooks/snapshots.ts:92]:
```suggestion
        snapshot.links.map(async (linkId) => {
```

**Review by wujessica — APPROVED** [review]

File: src/interfaces/coral_web/src/components/ShareModal.tsx (hunk 3 above)
Comment (misspia-cohere) [reply src/interfaces/coral_web/src/components/ShareModal.tsx]: ah good catch-- copied this from coral where the `linkId` could be null but that's not the case here

**Pull request merged**